### PR TITLE
Add codemirror keybindings and addons

### PR DESCRIPTION
**Pull request by DerMolly**
### Component/Part
markdown editor

### Description
This PR adds codemirror addons:
- fullScreen
- ~continueComment~ (this does not do anything in markdown documents)
- autorefresh
added a default for:
- extraKeys

### Steps

- [ ] added implementation
- [ ] added / updated tests
- [ ] added / updated documentation
- [ ] extended changelog

### CODE REVIEW

**Comment by DerMolly** [comment]
`continueComments` seems to be unnecessary

**Comment by DerMolly** [comment]
there seems to be a problem in key-map.ts currently, not sure how to fix that currently.

**Comment by ErikMichelson** [comment]
Is the fullscreen addon still not correctly working because of the way the codemirror editor is embedded? If so, I suggest to just drop that. Any user may switch to 'only edit'-view and may toggle the browser's fullscreen and have almost the same editor space.

**Comment by DerMolly** [comment]
I'm not sure if that's the way or if we should try to preserve the fullscreen mode. But yeah if we can't get it to work, removing it is the way…

**Comment by DerMolly** [comment]
I did just play around with that and the problem seems to be that the `CodeMirror` css class defines `position: relative;` and the fullscreen addon defines `CodeMirror-fullscreen` with `position: fixed;`. Both are applied to the same `<div>`. The question is is that a problem of `codemirror` (I somewhat doubt that, but could be) or `react-codemirror2`? Since the `CodeMirror` and `CodeMirror-fullscreen` classes both come from the codemirror npm package I'm a bit unsure how to fix this…

File: src/components/editor/editor-window/key-map.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  // },
+  // 'Shift-Ctrl-=': cm => {
+  //   utils.wrapTextWith(this.editor, cm, '+')
+  // },
+  // 'Ctrl-=': cm => {
+  //   utils.wrapTextWith(this.editor, cm, '=')
+  // },
+  // 'Shift-Ctrl-Backspace': cm => {
+  //   utils.wrapTextWith(this.editor, cm, 'Backspace')
+  // }
```
Comment (mrdrogdrog) [inline src/components/editor/editor-window/key-map.ts:110]: Is this code unneeded or not ready yet?

Comment (DerMolly) [reply src/components/editor/editor-window/key-map.ts:110]: not ready yet

Comment (DerMolly) [reply src/components/editor/editor-window/key-map.ts:110]: I'm currently debating to use the function `wrapTextWith` not only here but also for the toolbar as this seems to using the codemirror replace function directly which is hopefully better than the custom function over the whole content we're currently using. I will investigate more later and see if I can make it work…

File: src/components/editor/editor-window/key-map.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    wrapTextWith(cm, '^')
+  },
+  'Ctrl-+': cm => {
+    wrapTextWith(cm, '+')
+  },
+  'Shift-Ctrl-=': cm => {
+    wrapTextWith(cm, '+')
+  },
+  'Ctrl-=': cm => {
+    wrapTextWith(cm, '=')
```
Comment (ErikMichelson) [inline src/components/editor/editor-window/key-map.ts:107]: Why do you wrap the text with a single equality sign? This does nothing to the rendering. If you want a keybinding for hightlighting text, use double equality signs.

File: src/components/editor/editor-window/key-map.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      cm.setOption('fullScreen', false)
+    } else {
+      return CodeMirror.Pass
+    }
+  },
+  'Cmd-S': function () {
+    return CodeMirror.Pass
+  },
+  'Ctrl-S': function () {
+    return CodeMirror.Pass
```
Comment (ErikMichelson) [inline src/components/editor/editor-window/key-map.ts:21]: This just lets the keybinding pass forward to the browser and does not prevent the saving dialog. Same applies to the `Cmd-S` shortcut.

File: src/components/editor/editor-window/key-map.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
+import CodeMirror, { KeyMap } from 'codemirror'
+import { wrapTextWith } from './tool-bar/utils'
+
+// const isMac = CodeMirror.keyMap.default === CodeMirror.keyMap.macDefault
```
Comment (ErikMichelson) [inline src/components/editor/editor-window/key-map.ts:4]: Is this needed or can this safely be deleted?

File: src/components/editor/editor-window/key-map.ts (hunk 2 above)
Comment (DerMolly) [reply src/components/editor/editor-window/key-map.ts:107]: I just copied it from the old code

https://github.com/codimd/server/blob/master/public/js/lib/editor/index.js#L119-L121

, but I'll change it

File: src/components/editor/editor-window/key-map.ts (hunk 4 above)
Comment (DerMolly) [reply src/components/editor/editor-window/key-map.ts:4]: It was needed in the old code, but it's so stupid that I removed it, but I think we may need to readd this. Maybe we should rework the shortcuts altogether and remove the old ones for now?

Comment (mrdrogdrog) [reply src/components/editor/editor-window/key-map.ts:4]: Please let's try to build a clean solution

File: src/components/editor/editor-window/editor-window.tsx [hunk 5]
```diff
@@ -25,42 +28,15 @@ export interface EditorWindowProps {
...
     <div className={'d-flex flex-column h-100'}>
       <ToolBar
-        content={content}
-        onContentChange={onContentChange}
-        positions={positions}
+        editor={editor}
       />
       <ControlledCodeMirror
-        className="overflow-hidden w-100 flex-fill"
+        className="h-100 w-100 flex-fill"
```
Comment (mrdrogdrog) [inline src/components/editor/editor-window/editor-window.tsx:39]: Why did you remove the `overflow-hidden` ? It's important for the scrollbar

File: src/components/editor/editor-window/key-map.ts (hunk 4 above)
Comment (mrdrogdrog) [reply src/components/editor/editor-window/key-map.ts:4]: What was its purpose?

File: src/components/editor/editor-window/tool-bar/tool-bar.tsx [hunk 6]
```diff
@@ -1,82 +1,86 @@
...
-      </Button>
-      <Button variant='light' onClick={addLine} title={t('editor.editorToolbar.line')}>
-        <ForkAwesomeIcon icon="minus"/>
-      </Button>
-      <Button variant='light' onClick={addComment} title={t('editor.editorToolbar.comment')}>
-        <ForkAwesomeIcon icon="comment"/>
-      </Button>
-    </ButtonToolbar>
-  )
+  if (editor) {
```
Comment (mrdrogdrog) [inline src/components/editor/editor-window/tool-bar/tool-bar.tsx:34]: invert pls.

File: src/components/editor/editor-window/key-map.ts (hunk 3 above)
Comment (DerMolly) [reply src/components/editor/editor-window/key-map.ts:21]: fixed

File: src/components/editor/editor-window/key-map.ts (hunk 2 above)
Comment (DerMolly) [reply src/components/editor/editor-window/key-map.ts:107]: fixed

File: src/components/editor/editor-window/key-map.ts (hunk 4 above)
Comment (DerMolly) [reply src/components/editor/editor-window/key-map.ts:4]: it was used on some shortcuts to differentiate between macs and non-macs, but the latest changes should have made that obsolete

File: src/components/editor/editor-window/tool-bar/tool-bar.tsx (hunk 6 above)
Comment (DerMolly) [reply src/components/editor/editor-window/tool-bar/tool-bar.tsx:34]: done

File: src/components/editor/editor-window/editor-window.tsx (hunk 5 above)
Comment (DerMolly) [reply src/components/editor/editor-window/editor-window.tsx:39]: done

File: src/components/editor/editor-window/key-map.ts (hunk 1 above)
Comment (DerMolly) [reply src/components/editor/editor-window/key-map.ts]: As this is long gone code, could you close this @mrdrogdrog?

File: src/components/editor/editor-window/key-map.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
+import CodeMirror, { Editor, KeyMap, Pass } from 'codemirror'
+import {
+  makeSelectionBold,
+  makeSelectionItalic,
+  markSelection,
+  strikeThroughSelection,
+  underlineSelection
+} from './tool-bar/utils'
+
+const isMac = navigator.platform.includes('Mac')
```
Comment (mrdrogdrog) [inline src/components/editor/editor-window/key-map.ts:10]:
```suggestion
const isMac = navigator.platform.toLowerCase().includes('mac')
```
I think this wouldn't hurt. We never know if every browser will use "Mac".

File: src/components/editor/editor-window/key-map.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+const f10 = (editor: Editor): void | typeof Pass => editor.setOption('fullScreen', !editor.getOption('fullScreen'))
+const esc = (editor: Editor): void | typeof Pass => {
+  if (editor.getOption('fullScreen') && !isVim(editor.getOption('keyMap'))) {
+    editor.setOption('fullScreen', false)
+  } else {
+    return CodeMirror.Pass
+  }
+}
+const supressSave = (): undefined => undefined
```
Comment (mrdrogdrog) [inline src/components/editor/editor-window/key-map.ts:22]: sup<b>p</b>ress

File: src/components/editor/editor-window/key-map.ts (hunk 7 above)
Comment (DerMolly) [reply src/components/editor/editor-window/key-map.ts:10]: Well according to the Living HTML Standard the value should currently always be 'MacIntel'

File: src/components/editor/editor-window/key-map.ts (hunk 8 above)
Comment (DerMolly) [reply src/components/editor/editor-window/key-map.ts]: done

File: src/components/editor/editor-window/key-map.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  } else {
+    if (editor.getOption('indentWithTabs')) {
+      editor.execCommand('defaultTab')
+    } else {
+      editor.replaceSelection(spaces)
+    }
+  }
+}
+
+export const defaultKeyMap: KeyMap = isMac ? {
```
Comment (ErikMichelson) [inline src/components/editor/editor-window/key-map.ts:65]:
```suggestion
export const defaultKeyMap: KeyMap = !isMac ? {
```

The ternary condition needs to be inverted.
